File: section3/ValidParentheses.java
import java.util.Stack;

class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();

        for (char ch: s.toCharArray()) {
            if (stack.size() == 0 && (ch == ')' || ch == '}' || ch == ']')) {
                return false;
            }

            if (ch == '(' || ch == '{' || ch == '[') {
                stack.add(ch);
                continue;
            }

            char last = stack.lastElement();
            if ((last != '(' && ch == ')') || (last != '{' && ch == '}') || (last != '[' && ch == ']')) {
                return false;
            }
            stack.pop();
        }

        return stack.size() == 0 ? true : false;
    }
}

File: section3/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for ch in s:
            if ch is '(' or ch is '{' or ch is '[':
                stack.append(ch)
                continue

            top = stack[-1] if len(stack) > 0 else None
            if top is not '(' and ch is ')' or top is not '{' and ch is '}' or top is not '[' and ch is ']':
                return False
            stack.pop()
        
        return True if len(stack) == 0 else False       
